### ISSUE TITLE
Add non-interactive usage

### ISSUE THREAD

**Issue by bradwilson** [dn-vm/dnvm#147]
I use Ansible to automated [my development machines](https://github.com/bradwilson/ansible-dev-pc). Due to limitations imposed by Linux distributions, I am currently [using `dotnet-install`](https://github.com/bradwilson/ansible-dev-pc/blob/8af39a3c9dc30f869bb26e828e613c54989f7453/dotnet/core.yaml#L22-L32) to install specific SDKs and runtimes. This comes with some significant downsides; namely, there is no uninstallation and/or upgrade scenarios supported by `dotnet-install`. `dnvm` seems like a natural fit.

However, I cannot automated `dnvm` as it stands today because it mandates interactivity. To make the switch, I would require non-interactive invocations:

- Ability to install the tool (don't prompt me for anything, ideally just give me a fixed URL to grab the latest `dnvm` binary to drop into a `~/bin` folder)
- Ability to install an SDK
- Ability to install a runtime

I would of course use `dnvm` in its interactive mode to perform upgrades as necessary (in much the same way I use Ansible to get software initially installed when building the dev environment, but rely on `apt update && apt upgrade` to keep it up to date).

**Comment by richlander** [comment]
I just ran into this issue as well.

I was playing with Cursor Agents. It doesn't have .NET installed in its environment. It installed (the wrong) .NET via the install script. I thought I'd give it some AI-first instructions on how to use dnvm. It would be great if there was a non-interactive approach that works well in this scenario.

I'm happy to contribute to an AI-first README.

I'll probably start with AI instructions for the README. That will inform what we could do with dnvm.

Here's some of the chat log.

<img width="382" height="1295" alt="Image" src="https://github.com/user-attachments/assets/903ff0ab-6188-49a9-a6e9-b210391481c4" />

<img width="382" height="943" alt="Image" src="https://github.com/user-attachments/assets/00cdb943-677e-4b39-a4f0-732261ed930f" />

The feedback includes a reference to ChatGPT but I'm using Cursor. I asked ChatGPT for some peer-review feedback. It gave me good feedback and I shared that with Cursor as its marching orders.

**Comment by nagilson** [comment]
I believe `dnvm update` has `-y` to accept all by default, does that still require interactivity? Note that I tried to test myself but encountered an issue:

<img width="1103" height="211" alt="Image" src="https://github.com/user-attachments/assets/dc8589fa-12eb-421f-a19e-83b303a23969" />

**Comment by agocke** [comment]
The current status of this issue is that it’s about 50% implemented and 50% tested. 

As Noah mentioned, `-y` should basically do a standard install with no prompts, but it’s lightly tested. 

Another scenario is just dropping dnvm into a bin folder. That should also work, but it’s about 0% tested. The big difference there is that dnvm itself won’t live in DNVM_HOME, and I might have some places in the code that assume that.  

The last piece is a permanent latest url. This isn’t implemented. The trick is that I don’t run an actual webserver so I can’t do this with redirects. I have to deploy the binaries to the GitHub pages repo. I’m honestly not sure how well that will perform, but hopefully it’s good enough since dnvm is only about 5MB.

**Comment by bradwilson** [comment]
Even the stable URL is something I can work around by querying GitHub to know what the latest release is, as I use this technique for other downloads (like Docker Compose):

```shell
$ curl -s https://api.github.com/repos/dn-vm/dnvm/releases/latest | grep "browser_download_url.*linux-x64.tar.gz\"" | cut -d : -f 2,3 | tr -d \ \"
https://github.com/dn-vm/dnvm/releases/download/v0.9.7/dnvm-0.9.7-linux-x64.tar.gz
```

No idea if this helps @richlander's scenario for download or not. You can tweak the grep for the bits you're looking for:

<img width="1588" height="492" alt="Image" src="https://github.com/user-attachments/assets/d988f05b-165d-4c02-a4b2-a6d6f4994fe1" />

Conversion from Linux CLI tools to something else are an exercise left up to the reader. 😁 

-----

The big thing for me is being able to recreate the equivalent of having gone through the install, without interactivity. If that means setting know environment variables, that's something I could do. Better might be if we could repurpose some of the machinery inside to do a non-interactive version of the install, so that I'm not constantly tracking any changes you need to make to the shell startup in my Ansible script.

**Comment by agocke** [comment]
@bradwilson When you say "go through the install" -- do you want to just install dnvm or do you also want it to automatically install a .NET SDK? That is, obviously copying dnvm alone will only install dnvm. But running `curl ... | sh -s -y` will run with the default interactive options, which includes tracking the `latest` SDK. I'm curious which one would be better for you.

**Comment by bradwilson** [comment]
Yep, that works. https://github.com/bradwilson/ansible-dev-pc/blob/main/dotnet/core.yaml

**Comment by richlander** [comment]
> BOMs are a "Windows disease",

- https://github.com/dotnet/sdk/issues/39187
- https://github.com/dotnet/core/issues/9957

I appreciate your support in this fight.

**Comment by bradwilson** [comment]
@richlander I required all files to be BOM free automatically in my CI build script:

https://github.com/xunit/build-tools-v3/blob/3924f254f246ef8ac0eddf4a54f5b012faa74160/models/BuildContext.cs#L173-L204
https://github.com/xunit/build-tools-v3/blob/3924f254f246ef8ac0eddf4a54f5b012faa74160/targets/FormatSource.cs#L21-L31

And I removed BOMs from all my output reports (it was at request of a user, who wanted to post-process them, but I should've known better already): https://github.com/xunit/xunit/issues/3224

**Comment by bradwilson** [comment]
> I appreciate your support in this fight.

@richlander Did you want more than me just putting a thumbs up on the two issues? 😄

**Comment by agocke** [comment]
I'm doing my own json writing and it should definitely *not* be writing a BOM, so I have to investigate how that's getting in there.

**Comment by agocke** [comment]
Oh great, Encoding.UTF8 has a BOM by default...

**Comment by bradwilson** [comment]
> Oh great, Encoding.UTF8 has a BOM by default...

It sure is. 😂

**Comment by agocke** [comment]
OK, I've tested this workflow and it seems to work OK:

1. Set the DNVM_HOME env variable to point to your desired SDK install path (default ~/.local/share/dnvm)
2. Pick a place to install the dnvm exe. (default $DNVM_HOME/dnvm)
3. Run `curl --proto '=https' -sSf https://dnvm.net/install.sh | sh -s -- -y --dest-path <dnvm-path> --skip-tracking -f`

That will download the dnvm exe to the target path and set up the environment, but it won't track/install any SDKs. You can do that through subsequent calls to dnvm.

**Comment by bradwilson** [comment]
I'm going to make an argument that might be controversial, and technically I don't _need_ this, but I wanted to bring it up since we're talking about non-interactive usage and idempotency.

The hot take: **Calling `dnvm track <version>` when you already track that version shouldn't be an error condition**.

I understand why you made it be an error given the message you produce, but the fact that it's an error is what caused me to need to write the JSON parsing code (so I could remove anything that's already tracked from the list that I need to pass to `dnvm track`).

My argument is that I want to think of it like `Directory.CreateDirectory` in .NET. When a user calls `Directory.CreateDirectory` they just want to make sure the directory already exists, so why introduce a race condition by making me `Directory.Exists` first? This is presumably why they don't throw for an already existing directory.

Obviously the chance of a race condition here between the time I parse the JSON vs. the time I call `dnvm track` is effectively zero, so that's not a particularly good argument, but it does feel like I had to write the JSON parsing code unnecessarily.

I think questions like this come about when you contemplate what all the non-interactive usage scenarios are, and it would be entirely reasonable to behave differently interactively ("this feels wrong, did you mean this?") vs. non-interactively ("okay, we're already tracking that, so that's fine").
